pkg/operator/ceph/pool/status.go

```diff
@@ -0,0 +1,117 @@
+/*
+Copyright 2020 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package pool to manage a rook pool.
+package pool
+
+import (
+	"context"
+	"time"
+
+	cephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+	cephclient "github.com/rook/rook/pkg/daemon/ceph/client"
+	opcontroller "github.com/rook/rook/pkg/operator/ceph/controller"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// updateStatus updates a pool CR with the given status
+func updateStatus(client client.Client, poolName types.NamespacedName, status cephv1.ConditionType, info map[string]string) {
+	pool := &cephv1.CephBlockPool{}
+	err := client.Get(context.TODO(), poolName, pool)
+	if err != nil {
+		if kerrors.IsNotFound(err) {
+			logger.Debug("CephBlockPool resource not found. Ignoring since object must be deleted.")
+			return
+		}
+		logger.Warningf("failed to retrieve pool %q to update status to %q. %v", poolName, status, err)
+		return
+	}
+
+	if pool.Status == nil {
+		pool.Status = &cephv1.CephBlockPoolStatus{}
+	}
+
+	pool.Status.Phase = status
+	pool.Status.Info = info
+	if err := opcontroller.UpdateStatus(client, pool); err != nil {
+		logger.Warningf("failed to set pool %q status to %q. %v", pool.Name, status, err)
+		return
+	}
+	logger.Debugf("pool %q status updated to %q", poolName, status)
+}
+
+// updateStatusBucket updates an object with a given status
+func (c *mirrorChecker) updateStatusMirroring(mirrorStatus *cephclient.PoolMirroringStatus, mirrorInfo *cephclient.PoolMirroringInfo, details string) {
+	blockPool := &cephv1.CephBlockPool{}
+	if err := c.client.Get(context.TODO(), c.namespacedName, blockPool); err != nil {
+		if kerrors.IsNotFound(err) {
+			logger.Debug("CephBlockPool resource not found. Ignoring since object must be deleted.")
+			return
+		}
+		logger.Warningf("failed to retrieve ceph block pool %q to update mirroring status. %v", c.namespacedName.Name, err)
+		return
+	}
+	if blockPool.Status == nil {
+		blockPool.Status = &cephv1.CephBlockPoolStatus{}
+	}
+	blockPool.Status.MirroringStatus, blockPool.Status.MirroringInfo = toCustomResourceStatus(blockPool.Status.MirroringStatus, mirrorStatus, blockPool.Status.MirroringInfo, mirrorInfo, details)
```

Comment:
Does the mirroring status have a timestamp so it's clear how recent the check was?

Reply:
How about a last known good status? If the ceph calls start failing, it would be good to keep track of the last known status when we could connect.

Reply:
No timestamp coming from the status command. Ok for a last known status.

Reply:
Actually, on error, we don't change `Summary`, only the `details` which contains the `err.Error()`, so we will always know the previous status.

Reply:
Sounds good if we know the status. So you can add a timestamp of the summary?

Reply:
We already have a `lastChecked` value, isn't that enough?

Reply:
Will that be updated on both success and failure? It would be good to keep a separate timestamp for the last successful check even if things are failing. What if we only updated `lastChecked` on success? And maybe rename it `lastSuccessfulCheck`?

Reply:
We can update `lastChecked` on success only, renaming it would break the consistency with other status checks though...

Reply:
Ok sounds good, no need to rename